packages/ra-core/src/form/useInitializeFormWithRecord.ts

```diff
@@ -13,42 +13,9 @@ const useInitializeFormWithRecord = record => {
             return;
         }
 
-        const registeredFields = form.getRegisteredFields();
-
-        // react-final-form does not provide a way to set multiple values in one call.
-        // Using batch ensure we don't get rerenders until all our values are set
-        form.batch(() => {
-            Object.keys(record).forEach(key => {
-                // We have to check that the record key is actually registered as a field
-                // as some record keys may not have a matching input
-                if (registeredFields.some(field => field === key)) {
-                    if (Array.isArray(record[key])) {
-                        // array of values
-                        record[key].forEach((value, index) => {
-                            if (
-                                isObject(value) &&
-                                Object.keys(value).length > 0
-                            ) {
-                                // array of objects
-                                Object.keys(value).forEach(key2 => {
-                                    form.change(
-                                        `${key}[${index}].${key2}`,
-                                        value[key2]
-                                    );
-                                });
-                            } else {
-                                // array of scalar values
-                                form.change(`${key}[${index}]`, value);
-                            }
-                        });
-                    } else {
-                        // scalar value
-                        form.change(key, record[key]);
-                    }
-                    form.resetFieldState(key);
-                }
-            });
-        });
+        const initialValues = form.getState().initialValues;
+        const initialValuesMergedWithRecord = merge({}, initialValues, record);
+        form.initialize(initialValuesMergedWithRecord);
```

Comment:
This is neat but there is a difference with the previous code: you initialize values including the ones for which there is no associated input. You put that restriction in 3e56efe8caec8ab297dea9c3ccb8c293b5ff5309. Are you sure you're not reintroducing a bug?

Reply:
I did not find any

Reply:
Including the one from #3721?

Reply:
I'll add some tests tomorrow